addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyserUnitTest.java

```diff
@@ -0,0 +1,160 @@
+        for (HtmlContext ctx : contexts) {
+            Map<String, String> attrMap = ctx.getTagAttributes();
+            assertThat(attrMap.get("assert"), is(equalTo("x")));
+            assertThat(attrMap.get("onclick"), is(equalTo("alert('1')")));
+        }
+    }
+
+    @Test
+    void shouldGetAttributeName() throws Exception {
+        msg = new HttpMessage();
+        msg.setRequestHeader("GET /index.html HTTP/1.1");
+        msg.setResponseBody("<html><body assert=\"\"></body></html>");
+        HtmlContextAnalyser analyser = new HtmlContextAnalyser(msg);
+        String catcher = "assert";
+        List<HtmlContext> contexts = analyser.getHtmlContexts(catcher, null, 0);
+        assertThat(contexts.size(), is(equalTo(1)));
+        for (HtmlContext ctx : contexts) {
+            Map<String, String> attrMap = ctx.getTagAttributes();
+            assertThat(ctx.getParentTag(), is(equalTo("body")));
+            assertThat(attrMap.get("assert"), is(equalTo("")));
+        }
+    }
+
+    @Test
+    void shouldGetParsedAttributes() {
+        String attr = " onclick = \"alert(100)\" accesskey = \"x\" n=\"\"";
+        Map<String, String> attrMap = HtmlContextAnalyser.parseAttributes(attr);
+        assertThat(attrMap.get("onclick"), is(equalTo("alert(100)")));
+        assertThat(attrMap.get("accesskey"), is(equalTo("x")));
+        assertThat(attrMap.get("n"), is(equalTo("")));
+    }
+
+    @Test
+    void shouldGetParsedTags() {
+        String tag = "<body onclick = \"alert(100)\" accesskey = \"x\" n=\"\">";
+        Map<String, Map<String, String>> tagMap = HtmlContextAnalyser.parseTag(tag);
+        Map<String, String> attrMap = tagMap.get("body");
+        assertThat(attrMap.get("onclick"), is(equalTo("alert(100)")));
+        assertThat(tagMap.get("body").get("accesskey"), is(equalTo("x")));
+    }
+
+    @Test
+    void shouldParseTagsWithoutCompletion() {
+        String tag = "<html> <body> <p accesskey='x' onclick='alert(100)'> Hello </html>";
+        Map<String, Map<String, String>> tagMap = HtmlContextAnalyser.parseTag(tag);
+        Map<String, String> attrMap = tagMap.get("p");
+        assertThat(attrMap.get("onclick"), is(equalTo("alert(100)")));
+        assertThat(tagMap.get("p").get("accesskey"), is(equalTo("x")));
+    }
+
+    @Test
+    void shouldNotRaiseNullPointerException() {
+        String tag = "<html> <body> <p accesskey='x' onclick='alert(100)' Hello </html>";
+        Map<String, Map<String, String>> tagMap = HtmlContextAnalyser.parseTag(tag);
+        assertThat(tagMap.size(), is(equalTo(3)));
+        assertThat(tagMap.get("html").size(), is(equalTo(0)));
+        assertThat(tagMap.get("p").size(), is(equalTo(2)));
+        assertThat(tagMap.get("p").get("Hello"), is(equalTo(null)));
+    }
+
+    @Test
+    void shouldNotParseComments() {
+        String tag =
+                "<html> <body> <p accesskey='x' onclick='alert(100)'> Hello <!-- comment --> </html>";
```

Comment:
What happens if the p tag is no closed? Like in the above example.

Reply:
I have added a separate test for this condition too